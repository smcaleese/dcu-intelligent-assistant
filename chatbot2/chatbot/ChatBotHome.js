import React, {useState} from 'react';
import { StyleSheet, View, Alert } from 'react-native';
import CustomChatBot from '../components/CustomChatBot';
import AndroidKeyboardAdjust from 'react-native-android-keyboard-adjust';
import { useFocusEffect } from '@react-navigation/native';
import { getClasses } from '../utils/GetClasses';
import { getBusByNumber, getBusByStop, getStopID } from '../utils/GetBus';
import { searchMap } from '../utils/GetMapInfo';
import { checkToken } from '../utils/Users';
import { getAssignments } from '../utils/Assignments';
import AsyncStorage from '@react-native-async-storage/async-storage';

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: 'white'
    }
});

export default function ChatBotHome(props) {

    //Adjust keyboard when user accesses page
    useFocusEffect(() => {
        console.log("set keyboard to adjustResize")
        AndroidKeyboardAdjust.setAdjustResize()
    })

    // outputState preserves information in output so that it is still there 
    // when the "View Previos Output" button is pressed
    const [outputState, setOutputState] = useState({
        classesData: "no data",
        jsDate: ""
    })

    //Handle user intent for getting and displaying timetable data
    const handleTimeTable = (rawCourseCode, date, err) => {
        getClasses(rawCourseCode, date, err)
        .then((res) => {
            if(res === "server error") {
                console.log("server error: probably wrong course code")
                props.navigation.navigate("output", { classesData: "NA", jsDate: '' })
            } 
            else {
                const resData = {
                    classesData: res.classesData, 
                    dayOfTheWeek: res.dayOfTheWeek, 
                    jsDate: res.jsDate,
                    courseCode: res.courseCode
                }
                setOutputState(resData)
                props.setOutputTitle("Class Timetable")
                props.navigation.navigate("output", resData)
            }
        })
        .catch((err) => {
            console.log("server error:", err)
            props.navigation.navigate("output", { classesData: "NA", jsDate: '' })
        })
    }

    //Handle user intent for getting and displaying transport data by stop name/id/number
    const handleBusStop = (stop_data) => {
        getStopID(stop_data)
        .then(([stop_id, stop_name]) => {
            getBusByStop(stop_id)
            .then((res) => {
                console.log(res)
                console.log(stop_name)
                setOutputState({ times: res })
                props.setOutputTitle("Bus Timetable")
                props.navigation.navigate('bus', { times: res, input: stop_id, type: "stop"  })
            })
            .catch((err) => console.log(err))
        })
    }

    //Handle user intent for getting and displaying transpord data by bus number
    const handleBusNumber = (bus_number) => {
        getBusByNumber(bus_number)
        .then((res) => {
            setOutputState({ times: res })
            props.setOutputTitle("Bus Timetable")
            props.navigation.navigate('bus', { times: res, input: bus_number, type: "bus" })
        })
        .catch((err) => {
            console.log(err)
        })
    }

    //Handle user intent for getting and displaying map location
    const handleLocation = (location) => {
        searchMap(location)
        .then((res) => {
            props.setOutputTitle(res.title)
            setOutputState({ mapData: res })
            props.navigation.navigate('map', { mapData: res })
        })
        .catch((err) => {
            console.log(err)
        })
    }
    
    //Handle user intent for getting and displaying Assignments
    const handleRequestForAssignments = async () => {
        // see if user has logged in the past 24h
        const token = await AsyncStorage.getItem("@token")
        const [tokenIsValid, email] = await checkToken(token)
        if(tokenIsValid) {
            const assignments = await getAssignments(email)
            console.log("assignments:", assignments)
            props.setOutputTitle("Assignments")
            props.navigation.navigate('assignments', { assignments: assignments })
        }
        else {
            console.log("tokenIsValid:", tokenIsValid, "email:", email)
            Alert.alert("please login to access your assignments")
        }
    }

    //Identify intent and call the correct function when user request 'fulfilled' by chatbot
    const handleComplete = (err, confirmation) => {
        console.log("handleComplete")
        console.log("confirmation in ChatBotStack:", confirmation)
        
        switch(confirmation.intentName) {
            case("getClasses"):
                console.log("CASE 1: getClasses")
                handleTimeTable(confirmation["slots"].course, confirmation["slots"].date, err)
                break;
            case ("getBusByStop"):
                console.log("Case 2: getBusByStop")
                handleBusStop(confirmation["slots"].busstop)
                break;
            case ("getBusByNumber"):
                console.log("Case 3: getBusByNumber")
                handleBusNumber(confirmation["slots"].busnum)
                break;
            case ("getLocation"):
                console.log("Case 4: getLocation")
                handleLocation(confirmation["slots"].location)
                break;
            case ("getAssignments"):
                console.log("Case 5: getAssignments")
                handleRequestForAssignments()
        }
    }

    const goToOutput = () => {
        props.navigation.navigate("output", outputState)
    }
    
    return (
        <View style={styles.container}>
            <CustomChatBot
                style={styles.chatbot}
                goToOutput={goToOutput}
                handleComplete={handleComplete}
            />
        </View>
    )
}