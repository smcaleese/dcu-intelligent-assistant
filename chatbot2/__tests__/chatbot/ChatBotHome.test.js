import React from 'react';
import {create} from 'react-test-renderer';
import ChatBotHome from '../../chatbot/ChatBotHome';

jest.mock('@react-navigation/native');

const screen = create(<ChatBotHome />);

test('ChatBotHome Snapshot Test', () => {
    expect(screen).toMatchSnapshot();
})