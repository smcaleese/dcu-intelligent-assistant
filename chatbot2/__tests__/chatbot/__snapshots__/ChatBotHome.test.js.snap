// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatBotHome Snapshot Test 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <RCTScrollView
      onContentSizeChange={[Function]}
      style={
        Object {
          "flex": 1,
          "marginTop": 20,
        }
      }
    >
      <View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "marginBottom": 8,
                "marginTop": 8,
                "width": "100%",
              },
              Object {
                "justifyContent": "flex-start",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "borderRadius": 18,
                  "padding": 10,
                },
                Object {
                  "backgroundColor": "#c2c2c2",
                  "color": "black",
                  "marginLeft": 16,
                },
              ]
            }
          >
            Type 'help' for a list of example questions
          </Text>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "backgroundColor": "#dde",
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": 10,
          "width": "100%",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoFocus={true}
        onChangeText={[Function]}
        placeholder="Type here to talk"
        rejectResponderTermination={true}
        style={
          Object {
            "flex": 1,
            "fontSize": 16,
            "padding": 16,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "flex-end",
              "backgroundColor": "#f1f1f1",
              "height": 60,
              "justifyContent": "center",
              "opacity": 1,
              "width": 60,
            }
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={24}
            bbWidth={24}
            className=""
            fill="none"
            focusable={false}
            height={24}
            meetOrSlice={0}
            minX={0}
            minY={0}
            stroke="#000"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 24,
                  "width": 24,
                },
              ]
            }
            vbHeight={24}
            vbWidth={24}
            width={24}
          >
            <RNSVGGroup
              fill={null}
              propList={
                Array [
                  "fill",
                  "stroke",
                  "strokeWidth",
                  "strokeLinecap",
                  "strokeLinejoin",
                ]
              }
              stroke={4278190080}
              strokeLinecap={1}
              strokeLinejoin={1}
              strokeWidth={2}
            >
              <RNSVGPath
                d="M22 2L11 13M22 2l-7 20-4-9-9-4 20-7z"
              />
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
    <View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#f1f1f1",
            "borderRadius": 30,
            "margin": 16,
            "opacity": 1,
            "width": "50%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 16,
              "padding": 8,
              "textAlign": "center",
            }
          }
        >
          View Previous Output
        </Text>
      </View>
    </View>
  </View>
</View>
`;
